temp: Cross-platform BuildPulse GitHub Action

janpio opened the pull request:
cc @jasonrudolph 

As suggested in https://github.com/Workshop64/buildpulse-action/issues/2#issuecomment-1052684667

jasonrudolph committed:
Try out cross-platform BuildPulse GitHub Action

janpio commented on .github/workflows/test.yml:
```diff
@@ -723,16 +723,16 @@ jobs:
-          curl -fsSL https://get.buildpulse.io/test-reporter-darwin-amd64 > ./buildpulse-test-reporter
-          chmod +x ./buildpulse-test-reporter
-          ./buildpulse-test-reporter submit packages --account-id 17219288 --repository-id 192925833
+        if: startsWith(github.repository, 'prisma/') && && !cancelled()
```
```suggestion
        if: startsWith(github.repository, 'prisma/') && !cancelled()
```
@jasonrudolph

jasonrudolph replied:
Oops! Good catch. 🤦

jasonrudolph committed:
Fix syntax in `if` statement

Co-authored-by: Jan Piotrowski <piotrowski+github@gmail.com>

jasonrudolph commented:
Thanks for trying this out, @janpio. The good news is that the action is running for macOS and Windows ([screenshot](https://user-images.githubusercontent.com/2988/155863402-d898e35a-6651-4dec-9c37-d368b1850665.png)), but the bad news is that it's failing due to not having access to the secrets. I was expecting this `if` conditional to _prevent_ the step from running since this branch is from a fork:

https://github.com/prisma/prisma/blob/33fd05e7584daee77ca901a864b927882d8d47b8/.github/workflows/test.yml#L727

However, the step is still running.

I'll looking into this failure and let you know what I find out.

janpio commented on .github/workflows/test.yml:
```diff
@@ -723,16 +723,16 @@ jobs:
-          curl -fsSL https://get.buildpulse.io/test-reporter-darwin-amd64 > ./buildpulse-test-reporter
-          chmod +x ./buildpulse-test-reporter
-          ./buildpulse-test-reporter submit packages --account-id 17219288 --repository-id 192925833
+        if: startsWith(github.repository, 'prisma/') && !cancelled()
```
This could/should work maybe:
```suggestion
        if: github.event.pull_request.head.repo.full_name == github.repository && !cancelled()
```

janpio commented on .github/workflows/test.yml:
```diff
@@ -810,13 +810,12 @@ jobs:
-          curl -fsSL https://get.buildpulse.io/test-reporter-darwin-amd64 > ./buildpulse-test-reporter
-          chmod +x ./buildpulse-test-reporter
-          ./buildpulse-test-reporter submit packages/client --account-id 17219288 --repository-id 192925833
+        if: startsWith(github.repository, 'prisma/') && !cancelled()
```
This could/should work maybe:
```suggestion
        if: github.event.pull_request.head.repo.full_name == github.repository && !cancelled()
```

janpio reviewed (commented):
Good we caught that! Left 2 suggestions with code that could possibly work here (via https://github.community/t/how-to-detect-a-pull-request-from-a-fork/18363/3)

jasonrudolph committed:
Fix check for availability of secrets

If this works, it also has the benefit of being much more
intention-revealing. Prior to this change, we were checking the
organization name, which is a less explicit check.

jasonrudolph commented:
> Good we caught that! 

Indeed! :-)

> Left 2 suggestions with code that could possibly work here (via https://github.community/t/how-to-detect-a-pull-request-from-a-fork/18363/3)

Thanks. :bow: 

I gave that a try, but I discovered that it always returns false when there's no pull request. So any time there's a push with no pull request, it would skip the BuildPulse upload step. For example, when pushing to a branch before you have a pull request, it returns false and skips the step. I think the same thing would happen for the merge commit that's created when merging a pull request. That _seems_ undesirable to me, so I looked for another option.

Since we want to skip the step if secrets are unavailable, I looked to see if we could do that explicitly (as opposed to trying to check the organization name or the pull request info. At the moment, GitHub Actions doesn't support checking for secrets in a conditional (https://github.com/actions/runner/issues/1138), and it seems like [a lot of folks are running into this issue](https://github.com/actions/runner/issues/520). However, I came across [a solution that works in the meantime](https://github.com/jon-turney/meson/commit/af12c43843a8da6480487f4f42f10821399d0d82), and 8e92919d09d872091cb30ebc9971908da4b00067 adopts that approach. It's admittedly a bit of a hack, but I think it will meet our needs.

Since this branch is from a fork and therefore doesn't have access to the secrets, we can see that the BuildPulse step is skipped [here](https://github.com/prisma/prisma/runs/5347741810?check_suite_focus=true#step:20:50).

If the diff looks good to you, then I think the next step would be to add this code to a branch in this repository, and verify that the BuildPulse steps _do_ run.

@janpio: What do you think?

janpio commented:
Interesting. Feels like a pretty gaping hole in GH Actions then.

I'll do exactly that, merge this PR into a branch and then open a new PR for that one. Will @ mention you again.
